src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,252 @@
+
+        /// <summary>
+        /// Subscribes to the underlying <see cref="IObservable{T}"/> with the
+        /// transformation specified by this instance.
+        /// <br/><br/>
+        /// Disconnection requests via the returned <see cref="IDisposable"/> interface
+        /// are passed to the underlying <see cref="IObservable{T}"/> and also used
+        /// to signal pending asynchronous tasks that cancellation has been requested
+        /// and also used to prevent further event notifications.
+        /// </summary>
+        public IDisposable Subscribe(IObserver<TOut> observer)
+        {
+            IDisposable? subscription = null;
+            var newObserver = new Observer(observer, _transformNext, _transformError, () => subscription?.Dispose());
+            subscription = _source.Subscribe(newObserver);
+            return newObserver;
+        }
+
+        private class Observer : IObserver<TIn>, IDisposable
+        {
+            private CancellationTokenSource? _cancellationTokenSource = new();
+            private readonly CancellationToken _token;
+            //create a queue so that events will be sent in order
+            private readonly Queue<QueueEvent> _queue = new();
+            private readonly IObserver<TOut> _observer;
+            private readonly Func<TIn, CancellationToken, ValueTask<TOut>> _transformNext;
+            private readonly Func<Exception, CancellationToken, ValueTask<Exception>> _transformError;
+            private Action? _disposeAction;
+
+            public Observer(IObserver<TOut> observer, Func<TIn, CancellationToken, ValueTask<TOut>> transformNext, Func<Exception, CancellationToken, ValueTask<Exception>> transformError, Action disposeAction)
+            {
+                _token = _cancellationTokenSource.Token;
+                _observer = observer;
+                _disposeAction = disposeAction;
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<Exception>
+                _transformError = async (exception, token) => await transformError(exception, token).ConfigureAwait(false);
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<TOut>
+                _transformNext = async (data, token) => await transformNext(data, token).ConfigureAwait(false);
+            }
+
+            public void OnNext(TIn value) => Queue(QueueType.Data, _transformNext(value, _token), default);
+
+            public void OnError(Exception error) => Queue(QueueType.Error, default, _transformError(error, _token));
+
+            public void OnCompleted() => Queue(QueueType.Completion, default, default);
+
+            /// <summary>
+            /// Queues the specified event and if necessary starts watching for an event to complete.
+            /// </summary>
+            private void Queue(QueueType queueType, ValueTask<TOut> task, ValueTask<Exception> error)
+            {
```

Comment:
```suggestion
            {
                if (_token.IsCancellationRequested)
                    return;
```

Fast return to eliminate more possible race conditions in `Dispose` (after Cancel but before unsubscribe).

Reply:
It is not a race condition.  If the event is queued, it will be discarded.  I am actually going to make changes to the reverse; removing the short-circuit from ProcessDataAsync and ProcessErrorAsync to ensure that all Tasks are awaited eventually, even if the results are discarded.  If Tasks are not eventually awaited, weird things can happen as we had to fix in #2637

Reply:
Hmm, that was due to other code executing before it should.  In this case, well, ideally we should pause the disposal until the queue is empty, but of course pausing disposal has its own side-effects.  In this case I will go with your suggestion.

Reply:
```suggestion
            {
                if (_token.IsCancellationRequested)
                    return; // note: the task to be queued is never awaited in this case
```

Reply:
OK. This will lead to a change in design a bit. I'm done for today. I watched all code files, only two tests have left. If you make additional changes, then I will review them together with the 2 remaining tests.

Reply:
😄 Concurrency in comments detected.